src/nanoemoji/parts.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dataclasses
+from functools import lru_cache
+import json
+from nanoemoji.config import FontConfig
+from pathlib import Path
+from picosvg.svg import SVG
+from picosvg.svg_reuse import normalize
+from picosvg.svg_types import SVGPath, SVGShape
+from typing import Iterable, List, MutableMapping, Set, Tuple, Union
+
+
+PathSource = Union[SVGShape, Iterable[SVGShape], "ReuseableParts", Path]
+
+
+@lru_cache(maxsize=1)
+def _default_tolerence() -> float:
+    return FontConfig().reuse_tolerance
+
+
+def _is_iterable_of(thing, desired_type) -> bool:
+    try:
+        it = iter(thing)
+    except TypeError:
+        return False
+
+    try:
+        val = next(it)
+        return isinstance(val, desired_type)
+    except StopIteration:
+        return True
+
+
+@dataclasses.dataclass(frozen=True)
+class Shape:
+    d: str  # an SVG style path, e.g. the d attribute of <svg:path/>
+
+
+# A set of shapes that normalize to the same path
+# Only the d attribute of the SVGPath is populated
+@dataclasses.dataclass
+class ShapeSet:
+    normalized: Shape
+    shapes: Set[Shape] = dataclasses.field(default_factory=set)
+
+
+@dataclasses.dataclass
+class ReuseableParts:
+    version: Tuple[int, int, int] = (1, 0, 0)
+    reuse_tolerance: float = dataclasses.field(default_factory=_default_tolerence)
+    shape_sets: MutableMapping[Shape, ShapeSet] = dataclasses.field(
```

Comment:
isn't it redundant that the normalized shape is both the key of the shape_sets dictionary and an attribute of the `ShapeSet` class? is the latter needed, or used in absence of the key that is associated with the sets of shapes?
E.g. instead of `for shape_set in source.shape_sets.values():` (and then use `shape_set.normalized`) you could equally do:

`for normalized, shape_set in source.shape_sets.items():`

even the `ShapeSet` class feels redundant to me, you could reduce everything to a `MutableMapping[str, Set[str]]` just like your json representation basically

Reply:
I will see if I can use a NewType again. I'm not a fan of `Map<str, Set<str>>`, those types get very opaque very quickly.

Reply:
NewType works well. Brace yourself for my burst of NewType enthusisam leading to NewType's everywhere.